cmd/blsync, light/beacon: standalone beacon light sync tool

zsfelfoldi opened the pull request:
This PR implements a standalone tool called "beacon light syncer" that is capable of driving the engine API of an EL node while connected to an untrusted (possibly remote) CL node that supports the `light_client` namespace (Lodestar and Nimbus atm). While useful in itself, it is also the first step of delivering the PoS capable `LES/5` light client code. This PR contains the entire `SyncCommitteeTracker` logic, the SSZ binary merkle multiproof generator/verifier and a few beacon chain data structures.

The `blsync` tool can optionally also take advantage of a beacon state proof endpoint if available (in this case it can also set the finalized execution block root for the engine API). Note that such an endpoint is currently only provided by Lodestar and can be enabled by specifying the `--beacon.api.stateproof 1` parameter. There is also a [new endpoint format](https://github.com/ethereum/beacon-APIs/pull/267) proposed as a cross-client standard which is not released yet in any CL client but is already implemented in a currently unmerged Lodestar branch:
https://github.com/ChainSafe/lodestar/tree/cayman/proof-apis
The `blsync` tool is already prepared for working with the new API format when `--beacon.api.stateproof 2` is specified.

The tool can also perform a "dry run" (not drive any engine API when not specified) or run in test mode when the `--blsync.test` parameter is used. The test mode runs random beacon state requests while light syncing the beacon chain. Note that this mode only makes sense when the state proof endpoint is available and therefore can only be used with Lodestar.

I am running two CL nodes on the Sepolia chain with the `light_client` API publicly exposed for testing purposes: http://104.248.139.137:9596 (Lodestar, compiled from the new state proof branch) and http://104.248.139.137:5052 (Nimbus).

Some command line examples:
```
# Driving a local engine API from Nimbus CL:
./blsync --sepolia --beacon.api "http://104.248.139.137:5052" --blsync.engine.api "http://127.0.0.1:8551" --blsync.jwtsecret "jwt.hex"

# Driving a local engine API from Lodestar CL, including finality info from beacon state:
./blsync --sepolia --beacon.api "http://104.248.139.137:9596" --beacon.api.stateproof 2 --blsync.engine.api "http://127.0.0.1:8551" --blsync.jwtsecret "jwt.hex"

# Dry run with Nimbus CL:
./blsync --sepolia --beacon.api "http://104.248.139.137:5052"

# Test run with Lodestar CL:
./blsync --sepolia --beacon.api "http://104.248.139.137:9596" --beacon.api.stateproof 2 --blsync.test
```

Subsequent PRs will include:
- majority of the beacon chain structures and sync logic
- the historical stateRoots tree structure
- the LES/5 protocol additions and a refactoring of the protocol handler
- changes in the client logic for true "ultra-light" operation (header downloader no longer needed)

holiman commented on light/beacon/api/rest_api.go:
```diff
@@ -0,0 +1,340 @@
+		return nil, err
+	}
+
+	var beaconBlock struct {
```
neat construct. I did something less elegant: https://github.com/holiman/factor/blob/main/lib/types.go#L81

holiman commented on light/beacon/api/rest_api.go:
```diff
@@ -0,0 +1,357 @@
+
+// RestApi implements LightClientApi by requesting information from a beacon node REST API.
+// Note: all required API endpoints are currently only implemented by Lodestar.
+type RestApi struct {
+	Url string
+}
```
IMO the name `RestApi` does not properly describe what this thing is representing - I'm thinking that what this is representing, is a Remote CL-node. 
So I think you should rename it to `RemoteCL`, or `RemoteBeacon`.

holiman replied:
I also think you might extend it a bit, like I did here: https://github.com/holiman/factor/blob/main/lib/remote_cl.go#L30 

That means it can be used with custom headers. If we want to try these things with the nodes that the devops-team runs, we need to use special headers in order to gain access to the couldflare environment.

zsfelfoldi replied:
I named it `BeaconLightApi` because it's not always remote (in the blsync use case it typically is, but the same code will be used for LES servers that can and should probably use local beacon nodes).
I'll also do the custom header thing.

zsfelfoldi replied:
custon HTTP header is also done

zsfelfoldi commented on light/beacon/sync_committee.go:
```diff
@@ -0,0 +1,1684 @@
+	db                                                                                ethdb.KeyValueStore
+	sigVerifier                                                                       committeeSigVerifier
+	clock                                                                             mclock.Clock
+	bestUpdateCache, serializedCommitteeCache, syncCommitteeCache, committeeRootCache *lru.Cache
```
done

zsfelfoldi commented on light/beacon/sync_committee.go:
```diff
@@ -0,0 +1,1684 @@
+	GetInitData(ctx context.Context, checkpoint common.Hash) (Header, LightClientInitData, error)
+}
+
+// WeakSubjectivityCheckpoint implements SctConstraints in a way that it fixes the committee belonging to the checkpoint and allows
```
done

zsfelfoldi commented on light/beacon/sync_committee.go:
```diff
@@ -0,0 +1,1684 @@
+	return "", "", false
+}
+
+func LoadForks(fileName string) (Forks, error) { //TODO ignore in-line comments
```
It's the beacon chain config file (config.yaml), added comment

rjl493456442 commented:
A few general feedbacks:

### Beacon package

Let's put the beacon as a standalone level-0 package, namely it should be put in the root directory of codebase.
Inside of beacon package, we can have these subfolders like:
- engine: move core/beacon to this directory which defines the parameters in engine API
- bsync: the main directory of beacon-sync protocol
    - bsync/types: type definition like: checkpoint, forks, header
    - bsync/
- params: the protocol constants of beacon chain, e.g. the indices of beacon state fields
- merkletrie: the binary merkle tree implementation
- etc

### Link the specification

Would be huge appreciated if the corresponding specification can be linked at the type definitions.
For example, it's not easy to figure out what's the `CommitteeRoots` here, why it's a list of hash instead of a single root hash. And please annotate the fields since it's can be helpful to understand.

```go
// CheckpointData contains known committee roots based on a weak subjectivity checkpoint
type CheckpointData struct {
	Checkpoint     common.Hash
	Period         uint64
	CommitteeRoots []common.Hash
}
```

### Register blsync as a service

Currently blsync is a standalone command outside of Geth. I would vote to convert it as a service inside of Geth. This service can be started and stopped with the life cycle management by node stack. 

From another perspective, it's more convenient to run, maybe with a single flag `--blsync` is enough.

### Please remove the unused code

From the first round of reviewing, I found some codes are not used. I would vote to not include them in this PR. It will make review procedure definitely faster. For example, 

```go
// Activate allows broadcasting heads to the given peer
func (s *SyncCommitteeTracker) Activate(peer sctClient) {
	s.lock.Lock()
	defer s.lock.Unlock()

	s.broadcastTo[peer] = struct{}{}
	s.broadcastHeadsTo(peer, true)
}
```

### Code style

These suggestions are all from my own code style preferences, we can debt for sure :)

- Please ensure that the length of comment does not exceed 80 chars
- Please do not define multiple variables of the same type on the same line and please add descriptions to important variables/fields if possible
- Declare variable in block declaration format

For example, it's the original version

```go
// signingRoot calculates the signing root of the given header.
func (bf Forks) signingRoot(header Header) common.Hash {
	var signingRoot common.Hash
	hasher := sha256.New()
	headerHash := header.Hash()
	hasher.Write(headerHash[:])
	domain := bf.domain(uint64(header.Slot) >> 5)
	hasher.Write(domain[:])
	hasher.Sum(signingRoot[:0])
	return signingRoot
}
```

We can convert it to this format, it's will make eyes easier

```go
// signingRoot calculates the signing root of the given header.
func (forks Forks) signingRoot(header Header) common.Hash {
	var (
		signingRoot common.Hash
		headerHash  = header.Hash()
		hasher      = sha256.New()
		domain      = forks.domain(slotToEpoch(uint64(header.Slot)))
	)
	hasher.Write(headerHash[:])
	hasher.Write(domain[:])
	hasher.Sum(signingRoot[:0])
	return signingRoot
}
```

rjl493456442 commented on light/beacon/api/light_api.go:
```diff
@@ -0,0 +1,356 @@
+}
+
+// GetCheckpointData fetches and validates bootstrap data belonging to the given checkpoint.
+func (api *BeaconLightApi) GetCheckpointData(ctx context.Context, checkpoint common.Hash) (beacon.Header, beacon.CheckpointData, []byte, error) {
```
How can we verify the retrieved checkpoint is matched with the provided checkpointHash?

We only verify the retrieved committee is indeed matched with the associated header in the RPC response, but the entire response is not verified with the provided checkpointHash?

zsfelfoldi replied:
Indeed, good catch, I forgot to add this check.

rjl493456442 commented on light/beacon/api/light_api.go:
```diff
@@ -0,0 +1,354 @@
+	for k, v := range api.customHeaders {
+		req.Header.Set(k, v)
+	}
+	if resp, err := api.client.Do(req); err == nil {
```
```go
resp, err := api.client.Do(req)
if err != nil {
	return nil, err
}
defer resp.Body.Close()
return io.ReadAll(resp.Body)
```

rjl493456442 commented on light/beacon/api/light_api.go:
```diff
@@ -0,0 +1,354 @@
+// validity depends on the update chain.
+func (api *BeaconLightApi) GetBestUpdateAndCommittee(period uint64) (beacon.LightClientUpdate, []byte, error) {
+	periodStr := strconv.Itoa(int(period))
+	resp, err := api.httpGet("/eth/v1/beacon/light_client/updates?start_period=" + periodStr + "&count=1")
```
we can inline the periodStr though

```suggestion
	resp, err := api.httpGet("/eth/v1/beacon/light_client/updates?start_period=" + strconv.Itoa(int(period)) + "&count=1")
```

rjl493456442 commented on light/beacon/api/light_api.go:
```diff
@@ -0,0 +1,354 @@
+	if err := update.Validate(); err != nil {
+		return beacon.LightClientUpdate{}, nil, err
+	}
+	if beacon.SerializedCommitteeRoot(committee) != update.NextSyncCommitteeRoot {
```
It's duplicated. `update.NextSyncCommitteeRoot` is derived from committee

rjl493456442 commented on light/beacon/beacon_chain.go:
```diff
@@ -0,0 +1,102 @@
+var BsiFinalExecHash = ChildIndex(ChildIndex(BsiFinalBlock, BhiStateRoot), BsiExecHead)
+
+// Header defined a beacon header and supports JSON encoding according to the standard beacon API format
+type Header struct {
```
We can define a bunch of functions of Header, like `epoch`, `period` etc, it will make code looks like more readable.

zsfelfoldi replied:
I defined Epoch and SyncPeriod (also added a separate json struct to the api package because the `common.Decimal` <-> `uint64` conversions everywhere were really ugly)

rjl493456442 commented on light/beacon/sync_committee.go:
```diff
@@ -0,0 +1,547 @@
+
+// Validate verifies the validity of the update
+func (update *LightClientUpdate) Validate() error {
+	if update.Header.Slot&0x1fff == 0x1fff {
```
We can define the function of header `lastSlotOfPeriod`, it will make code more readable.

zsfelfoldi replied:
Actually this whole solution was ugly, now it's not needed at all because I pass the assumed `signatureSlot` separately to the sig verifier so I can just assume that updates are always signed by the same committee that the header belongs to (that is a validity condition of the header, the sig verification will fail if it's not true).

rjl493456442 commented on light/beacon/sync_committee.go:
```diff
@@ -0,0 +1,547 @@
+// Validate verifies the validity of the update
+func (update *LightClientUpdate) Validate() error {
+	if update.Header.Slot&0x1fff == 0x1fff {
+		// last slot of each period is not suitable for an update because it is signed by the next period's sync committee, proves the same committee it is signed by
```
Please link the [spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/validator.md#sync-committee), it's easier to understand why the last slot of period is not considered with update.

rjl493456442 commented on light/beacon/sync_committee.go:
```diff
@@ -0,0 +1,547 @@
+func (update *LightClientUpdate) Validate() error {
+	if update.Header.Slot&0x1fff == 0x1fff {
+		// last slot of each period is not suitable for an update because it is signed by the next period's sync committee, proves the same committee it is signed by
+		return errors.New("Last slot of period")
```
Why the last slot of period is always rejected? 

It can also be signed by the committee of next period which should be aware by light client already.

zsfelfoldi replied:
The client is not aware of the next period committee when validating this proof since this is the proof that should prove it to the client. It proves nothing if it's signed by the same committee.

rjl493456442 commented on light/beacon/sync_committee.go:
```diff
@@ -0,0 +1,547 @@
+			return errors.New("FinalizedHeader is from previous period") // proves the same committee it is signed by
+		}
+		if root, ok := VerifySingleProof(update.FinalityBranch, BsiFinalBlock, MerkleValue(update.FinalizedHeader.Hash()), 0); !ok || root != update.Header.StateRoot {
+			return errors.New("Invalid FinalizedHeader merkle proof")
```
Error string should not be capitalized or end with punctuation mark

rjl493456442 commented on light/beacon/sync_committee.go:
```diff
@@ -0,0 +1,547 @@
+	}
+	if update.hasFinalizedHeader() {
+		if update.FinalizedHeader.Slot>>13 != update.Header.Slot>>13 {
+			return errors.New("FinalizedHeader is from previous period") // proves the same committee it is signed by
```
Error string should not be capitalized or end with punctuation mark

rjl493456442 commented on light/beacon/sync_committee_head.go:
```diff
@@ -0,0 +1,304 @@
+// AllowBroadcast should be called when given head can be broadcast to other peers. Depending on the setup further data retrieval and/or
+// processing might be needed for a new head before it is propagated further.
+// Note that a signed head is broadcast when both addSignedHeads and AllowBroadcast has been called for the given head (in any order).
+func (s *SyncCommitteeTracker) AllowBroadcast(hash common.Hash) {
```
The chain head broadcast mechanism is not used in this PR. I know it might be used in the following PR, but I will vote to not include these logics in this PR to minimize the PR size.

zsfelfoldi replied:
Indeed, I removed most of the unused code but not this one. Done now.

rjl493456442 commented on light/beacon/sync_committee_update.go:
```diff
@@ -0,0 +1,545 @@
+
+// betterThan returns true if update u is considered better than w.
+func (u *UpdateScore) betterThan(w UpdateScore) bool {
+	uFinalized, wFinalized := u.isFinalized(), w.isFinalized()
```
From the spec, the comparison function is also defined, https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md#is_better_update

I feel like we should strictly follow the spec just re-implement it in golang.

zsfelfoldi replied:
I thought about it a bit more, also discussed it with @wemeetagain and we agreed that we should try to change the spec. I opened a PR against the specs repo where I also explained the reasons:
https://github.com/ethereum/consensus-specs/pull/3128/
(it is the same algorithm as implemented in this PR)

MariusVanDerWijden reviewed (commented):
Only managed to go through the first couple files, some nits. 
Need a proper pc for an real review

rjl493456442 commented on light/beacon/api/light_api.go:
```diff
@@ -0,0 +1,372 @@
+		return beacon.LightClientUpdate{}, nil, errors.New("invalid number of pubkeys in next_sync_committee")
+	}
+
+	committee, ok := c.NextSyncCommittee.serialize()
```
We should also check if the responded update is matched with period.

It should be very simple

`c.Header.Period  == period`

zsfelfoldi replied:
done

rjl493456442 commented on light/beacon/api/syncer.go:
```diff
@@ -0,0 +1,247 @@
+	checkpointCommittee []byte
+	sct                 *beacon.SyncCommitteeTracker
+
+	updateCache, committeeCache *lru.Cache
```
Let's use the new generic cache instead!

zsfelfoldi replied:
done

rjl493456442 commented on light/beacon/api/syncer.go:
```diff
@@ -0,0 +1,247 @@
+	genesisData         beacon.GenesisData
+	checkpointPeriod    uint64
+	checkpointCommittee []byte
+	sct                 *beacon.SyncCommitteeTracker
```
I will vote something like this. `sct` is just to hard to figure what it is.

zsfelfoldi replied:
done

rjl493456442 commented on light/beacon/api/syncer.go:
```diff
@@ -0,0 +1,247 @@
+func (cs *CommitteeSyncer) headPollLoop() {
+	timer := time.NewTimer(0)
+	var (
+		counter      int
```
This counter is never be increased.

zsfelfoldi replied:
Resolving this because `headPollLoop` has been rewritten (see above).

rjl493456442 commented on light/beacon/api/syncer.go:
```diff
@@ -0,0 +1,247 @@
+					timerStarted = true
+				}
+			}
+		case _, ok := <-cs.headTriggerCh:
```
Same feeling.

zsfelfoldi replied:
The `headPollLoop` was in a bit weird state, now it's rewritten and hopefully makes sense. Still, it's designed not just for this use case but also for light server mode where Geth is notified about new heads through the engine API. In this case it makes sense to trigger the polling externally when a new head is received and not poll the API endpoint all the time. For this purpose there is a `TriggerNewHead` function that triggers `headTriggerCh`. This function is not part of this PR because we agreed that currently unused functions should not be included in it.

rjl493456442 commented on light/beacon/api/light_api.go:
```diff
@@ -0,0 +1,372 @@
+		NextSyncCommitteeBranch beacon.MerkleValues `json:"next_sync_committee_branch"`
+		FinalizedHeader         jsonHeader          `json:"finalized_header"`
+		FinalityBranch          beacon.MerkleValues `json:"finality_branch"`
+		Aggregate               syncAggregate       `json:"sync_aggregate"`
```
Please add SignatureSlot according to the latest spec

zsfelfoldi replied:
Indeed, maybe it's better to decode and check it explicitly. Though later the signature period is assumed to be the same and the sig check fails if it is not, but now we catch it earlier and return a more informative error.

zsfelfoldi commented on cmd/blsync/main.go:
```diff
@@ -0,0 +1,240 @@
+
+	chainConfig := makeChainConfig(ctx)
+
+	if !ctx.IsSet(utils.BeaconApiFlag.Name) {
```
done

zsfelfoldi commented on cmd/blsync/config.go:
```diff
@@ -0,0 +1,197 @@
+	if customConfig && config.Forks != nil {
+		utils.Fatalf("Cannot use custom beacon chain config flags in combination with pre-defined network config")
+	}
+	if ctx.IsSet(utils.BeaconConfigFlag.Name) {
+		if forks, err := beacon.LoadForks(ctx.String(utils.BeaconConfigFlag.Name)); err == nil {
+			config.Forks = forks
+		} else {
+			utils.Fatalf("Could not load beacon chain config file", "file name", ctx.String(utils.BeaconConfigFlag.Name), "error", err)
+		}
+	}
```
done

zsfelfoldi commented on cmd/blsync/main.go:
```diff
@@ -0,0 +1,240 @@
+	var resp cbeacon.PayloadStatusV1
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	err := client.CallContext(ctx, &resp, "engine_newPayloadV1", *cbeacon.BlockToExecutableData(block))
+	cancel()
```
why would that be better?

NorutoSDK replied:
What happen

zsfelfoldi commented on cmd/blsync/config.go:
```diff
@@ -0,0 +1,197 @@
+	return cl
+}
+
+func obtainJWTSecret(fileName string) (jwtSecret [32]byte) {
```
done

holiman commented on light/beacon/sync_committee_update.go:
```diff
@@ -0,0 +1,544 @@
+	return u.signerCount+w.reorgRiskPenalty() > w.signerCount+u.reorgRiskPenalty()
+}
+
+// Encode encodes the update score data in a compact form for in-protocol advertisement
+func (u *UpdateScore) Encode(data []byte) {
+	v := u.signerCount + u.subPeriodIndex<<10
+	if u.finalizedHeader {
+		v += 0x800000
+	}
+	var enc [4]byte
+	binary.LittleEndian.PutUint32(enc[:], v)
+	copy(data, enc[:3])
+}
```
Where is the encoding format described?

zsfelfoldi replied:
Now it's described in the `UpdateScore` description.

fjl commented:
@rjl493456442 

> Currently blsync is a standalone command outside of Geth. I would vote to convert it as a service inside of Geth. This service can be started and stopped with the life cycle management by node stack.

Regarding this, we discussed to add it a bit later. It's very nice to have `blsync` as a standalone tool because there is lots of new code which could crash geth. We can also test it with other EL implementations to ensure the thing works with the standard engine API.

zsfelfoldi commented on light/beacon/api/syncer.go:
```diff
@@ -0,0 +1,247 @@
+		select {
+		case <-timer.C:
+			timerStarted = false
+			if head, err := cs.api.GetHeadUpdate(); err == nil {
```
Resolving this because `headPollLoop` has been rewritten (see below).

zsfelfoldi commented on light/beacon/sync_committee_signature.go:
```diff
@@ -0,0 +1,266 @@
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/minio/sha256-simd"
+	bls "github.com/protolambda/bls12-381-util"
```
I could not figure out yet how to verify a signature with `crypto.bls12381` if it's even possible, it looks to me like some necessary logic is missing from that package. I'm fine with using it though if there is a solution or we can add the missing logic there.

zsfelfoldi replied:
Recently we discussed this with @fjl and concluded that `crypto.bls12381` should be removed and `github.com/kilic/bls12-381` should be used instead as the former is an incomplete and outdated package and mostly an old fork of the latter anyway. The same library is used by proto's frontend which I am using here so this reference can stay as it is and I am going to open a separate PR that removes `crypto.bls12381` and changes the references to the external package (which can and should of course be tested separately). So eventually we're going to use the same bls library.

zsfelfoldi commented on cmd/blsync/main.go:
```diff
@@ -0,0 +1,237 @@
+			break
+		}
+		execBlockRoot := common.Hash(proof.Values[stateIndexMap[beacon.BsiExecHead]])
+		finalizedBeaconRoot := common.Hash(proof.Values[stateIndexMap[beacon.BsiFinalBlock]])
```
It's unique per proof format. The `MultiProof` does not have a `stateIndexMap` internally so it's probably better to keep this thing external but I changed it a bit to make it more readable (now I pre-calculate and store the actual value indices instead of the map) and also added proper comments to explain the meaning of the index variables.

zsfelfoldi commented:
@rjl493456442 I changed to block variable declarations in many places, in some places it obviously made the code look nicer, in some cases I wasn't totally sure, so please PTAL, I'm open to more suggestions. I also reformatted the comments and added links to the data structure specs wherever it made sense.
Unused code has also been removed, the last thing I'm still working on is the package structure refactoring.

rjl493456442 commented on light/beacon/beacon_chain.go:
```diff
@@ -0,0 +1,125 @@
+// https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#beaconblockheader
+type Header struct {
+	Slot          uint64
+	ProposerIndex uint
```
Nitpick, let's define it as uint64 according to spec.
`ValidatorIndex | uint64 | a validator registry index`

zsfelfoldi replied:
done

rjl493456442 commented on light/beacon/binary_merkle_test.go:
```diff
@@ -0,0 +1,264 @@
+		return true
+	}
+	if f1 == nil || f2 == nil {
+		return true
```
return false

zsfelfoldi replied:
Indeed, and also facepalm :) Turns out one of the tests was actually wrong, failed after fixing this return value. It's also fixed now.

rjl493456442 commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+
+// ProofWriter allow collecting data for a partial proof while a subset of a tree is traversed.
+type ProofWriter interface {
+	children() (left, right ProofWriter) // all non-nil subtrees are traversed
```
I think we should return an extra indicator if the proof is exhausted.
The nil children feels unintuitive.

zsfelfoldi replied:
There is no such thing as "exhausted" here, this interface traverses a certain shape, passes data in the process but does not care about where that data comes from or where it goes. A `nil` child means that it does not want to traverse any deeper from that given node (hence no ProofWriter for its children). So I feel it is appropriate and intuitive to pass `nil` here.

rjl493456442 commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+	if lzi == lzr {
+		if rf.subtree != nil && rf.index >= rf.begin && rf.index <= rf.end {
+			if st := rf.subtree(rf.index); st != nil {
+				return st.children()
```
If current cursor is as the same level with `begin` and `end`, subTree can only traverse into the node pointed by `index`. How can index be moved to next position?

rjl493456442 commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+		i1 = rf.index << (lzi - lzr)
+		i2 = ((rf.index + 1) << (lzi - lzr)) - 1
+	)
+	if i1 <= rf.end && i2 >= rf.begin {
```
For example, begin=8, and end=13; then i1=8, i2=15.

This condition is weird. `i1 <= rf.end && i2>=rf.begin` isn't always true unless index crosses this level where begin and end locates?

zsfelfoldi replied:
It becomes false when no descendants of `rf.index` are in the `begin..end` range. In your example, for `rf.index == 7` it's false so index 7 will be stored in the proof but 14 and 15 won't be.
I added some explanatory commments to the code because it's probably not trivial at first glance to see what is happening.

rjl493456442 commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+// level 0) and if greater than len(proof)-1 then the sibling subtrees closest
+// to the root are not stored and all-zero value subtrees are assumed (makes
+// proofs for 24 bit arrays more efficient).
+func VerifySingleProof(proof MerkleValues, index uint64, value MerkleValue, bottomLevel int) (common.Hash, bool) {
```
I am wondering if `bottomLevel` is actually useful. 

Usually if we ask something from fullNode, it will respond a value with attached proof. Will this proof omit zero-values by default?
If no, the zero-values is transmitted over p2p anyway and we can just use them.

zsfelfoldi replied:
Indeed, it's not used anywhere now so I removed it. Originally I wanted to optimize historical block proofs in LES but then I realized it's unnecessary overcomplication to use a special proof format in order to save a few hundred bytes. In the latest version I'm using the same multiproof format for all beacon state proofs and single proofs are only used in bootstrap data and committee updates (where `bottomLevel` is not needed).

rjl493456442 commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+// hash is not available then subtrees are always traversed (first left, then right).
+// If internal hash is available then subtrees are only traversed if needed by the writer.
+type ProofReader interface {
+	children() (left, right ProofReader) // subtrees accessible if not nil
```
I think we should return an extra indicator if the proof is exhausted.
The nil children feels unintuitive.

zsfelfoldi replied:
See above.

rjl493456442 commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+// ParseSSZMultiProof creates a MultiProof from a standard serialized format
+// (2 byte subtree sizes, then 32 byte leaf values and internal sibling hashes
+// in depth-first, left-to-right order).
+func ParseSSZMultiProof(proof []byte) (MultiProof, error) {
```
Can you elaborate the format of passed in multiProof?
Looks like it has
(1) a single byte prefix, maybe version?
(2) 2 bytes slice, maybe indices?
(3) 32 bytes slice, maybe leaves+internal nodes?

zsfelfoldi replied:
I added a detailed description.

rjl493456442 commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+// readNode implements ProofReader
+func (mpr multiProofReader) readNode() (MerkleValue, bool) {
+	if l, _ := mpr.format.children(); l == nil && len(*mpr.values) > 0 {
+		hash := (*mpr.values)[0]
```
Maybe we can maintain a pointer inside of multiProofReader.
pointer = len(values) if it's exhausted.

zsfelfoldi replied:
I did the change, though the pointer still needs to be a `*int` reference for the same reason as in the case of `multiProofWriter`. Even so, it probably looks nicer than devouring the slice from its head :)

rjl493456442 commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+func (mpw multiProofWriter) writeNode(node MerkleValue) {
+	if mpw.values != nil {
+		if lf, _ := mpw.format.children(); lf == nil {
+			*mpw.values = append(*mpw.values, node)
```
Why not define function with `*multiProofWriter` as the function receiver, so that you can directly append value into `values`.

zsfelfoldi replied:
Even if we could modify our `multiProofWriter` instance, that would not help us. The slice needs to be kept as a reference because the child writers append to the same list as the tree is traversed. If we copied the slice and the children appended to the copied slice, the new elements would not appear in the parent instance. I added explanatory comment to the code.

zsfelfoldi commented:
While rebasing to latest master I also squashed a lot of commits that seemed to have resolved previous issues a while ago, in order to avoid total chaos and painful rebase process. I think it's going to be easier to focus on more recent changes now but if someone wants to check out the individual steps of previous changes I saved the old commit history on this branch:
https://github.com/zsfelfoldi/go-ethereum/tree/blsync-before-squash

zsfelfoldi commented:
Also note that I moved `core/beacon` into `beacon/engine` according to Gary's suggestion. This touches many other parts of the codebase, `beacon.` prefixes have also been replaced to `engine.`. I think it's nice because it's basically 100% related to the engine api and now we have a top level `beacon` package so it looks good as a subpackage of that (it's also something we do that's related to the beacon chain).
I separated the changes affecting the existing codebase in a separate commit so that it can either be dropped easily if it gets voted down or I can also easily put it in a separate PR if that's better.

zsfelfoldi commented on light/beacon/binary_merkle.go:
```diff
@@ -0,0 +1,555 @@
+		}
+		hasher.Reset()
+		if index&1 == 0 {
+			hasher.Write(value[:])
```
`value` is a `merkle.Value` which is not a slice but a `[32]byte` so it's passed by value (it's copied already) and has a guaranteed size. So I think it should be ok.

holiman commented:
`--beacon.api <beacon api> --blsync.api <el node>` -- wouldn't `--cl.api` and `--el.api` be better? or `blsync.cl-api`, `blsync.el-api`

rjl493456442 commented on cmd/blsync/main.go:
```diff
@@ -0,0 +1,243 @@
+	if blocks == nil {
+		return
+	}
+	if !parentFound && maxBlockFetch == 0 {
```
Seems like the only purpose of this loop is to collect as many beacon-execution root as possible in the `e.execRootCache`. Is it really necessary? Personally I agree with martin that only the latest block is required for triggering sync. 

EL will handle reorg(in downloader) in case the provided new head header is not in the same chain as the previous local head.

zsfelfoldi replied:
This second loop is not needed for triggering sync (indeed, for that the head block would be enough). Collecting `execRootCache` associations is a lot cheaper than fetching entire blocks and it is used to serve the finalized execution block correctly. From the beacon state I get the finalized consensus block root and also the current consensus block root -> exec block root association but the finalized root points to an older block, therefore I need to collect older consensus -> exec root associations in order to know the finalized exec block root (which is I believe a nice thing to have in Geth and it does not cost a lot of resources).

zsfelfoldi replied:
I simplified this part as we discussed earlier. Now the beaconRoot -> execRoot associations are fetched by a separate function (so that it does not clutter up `newHead`). Also, due to popular demand, I removed the part that fetched multiple exec blocks and fed them into the engine API :) Now only the head block is fetched and the EL client can take care of the rest.

holiman commented on cmd/blsync/main.go:
```diff
@@ -0,0 +1,243 @@
+		}
+	}
+	for i := len(blocks) - 1; i >= 0; i-- {
+		blockRoot := blocks[i].Hash()
```
On a completely separate note... `blocks[i]` is `*ctypes.Block`. If the `ctype.Block` method `Hash()` returns the _stateRoot_ as opposed to the blockhash, then I think that is bad API, and not what I would expect from somthing called "block".

zsfelfoldi replied:
I don't follow. State root was never mentioned here, these are all block roots. `ctypes` is `core/types` so we are talking about good old execution blocks here and their `Hash()` definitely returns the block root. The beacon header's `Hash()` function does the same of course, state root is accessible as a field in both cases.

holiman commented:
Triage discussion: we should schedule a review call for this

MariusVanDerWijden commented on beacon/engine/types.go:
```diff
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU Lesser General Public License
 // along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
 
-package beacon
```
Moving `core/beacon` to `beacon/engine` kinda warrants the question whether we also want to move `eth/catalyst/` to something more descriptive, maybe `beacon/eth` or `eth/engine` (same with les/catalyst)

holiman commented on beacon/light/api/light_api.go:
```diff
@@ -0,0 +1,680 @@
+	}
+	encFormat, bitLength := EncodeCompactProofFormat(format)
+	if api.stateProofVersion >= 2 {
+		_, status, err := api.httpGet("/eth/v0/beacon/proof/subscribe/states?format=0x" + hex.EncodeToString(encFormat) + "&first=" + strconv.Itoa(first) + "&period=" + strconv.Itoa(period))
```
```suggestion
	_, status, err := api.httpGet("/eth/v0/beacon/proof/subscribe/states?format=#%x&first=%d&period=%d, ",
		encFormat, first, period)
```

... and that suggestion applies to a lot of places in this file. I don't quite get it why you are concatenating strings like that, it's not a common thing to do in go-lang.. ?

holiman replied:
You could even modify `api.httpGet` to take arguments like `Printf`.. e.g  `httpGetf(format string, a ...any)` and do the interpolation there

MariusVanDerWijden commented on beacon/light/api/light_api.go:
```diff
@@ -0,0 +1,680 @@
+	"strconv"
+	"time"
+
+	"github.com/donovanhide/eventsource"
```
Eventsource seems pretty unmaintained: https://github.com/donovanhide/eventsource
No commits in the last two years, failing ci, old issues etc.
Maybe it makes sense to use a different lib

zsfelfoldi commented:
Replaced by https://github.com/ethereum/go-ethereum/pull/26874
